File: scripts/automation.ts

import "dotenv/config";
import fs from "fs-extra";
import path from "path";
import { spawn, spawnSync } from "child_process";
import { loadConfig } from "../src/config.js";

const ROOT = path.resolve(path.dirname(new URL(import.meta.url).pathname), "..");

function logPath(base: string, override: string | undefined, name: string): string {
    const dir = override ? path.resolve(override) : path.resolve(base);
    return path.join(dir, `${name}.log`);
}

async function startServices(automation: any) {
    const logDir = automation.log_dir || "logs/automation";

    for (const service of (automation.services || [])) {
        if (service.enabled === false) continue;

        const cmdList = service.command || [];
        if (cmdList.length === 0) continue;

        const cwd = service.cwd ? path.resolve(service.cwd) : ROOT;
        const env = { ...process.env, ...(service.env || {}) };
        const name = service.name || "service";
        const logFile = service.log_file;
        const background = service.background || false;

        const lp = logPath(logDir, logFile, name);
        fs.ensureDirSync(path.dirname(lp));

        const out = fs.openSync(lp, "a");
        const err = fs.openSync(lp, "a");

        console.log(`Starting service: ${name}`);

        if (background) {
            const child = spawn(cmdList[0], cmdList.slice(1), {
                cwd,
                env,
                stdio: ['ignore', out, err],
                detached: true
            });
            child.unref();
        } else {
            spawnSync(cmdList[0], cmdList.slice(1), {
                cwd,
                env,
                stdio: ['ignore', out, err]
            });
        }
    }
}

function buildSchedule(automation: any): string[] {
    const lines: string[] = [];
    const logDir = automation.log_dir || "logs/automation";

    for (const schedule of (automation.schedules || [])) {
        if (schedule.enabled === false) continue;

        const cwd = schedule.cwd ? path.resolve(schedule.cwd) : ROOT;
        const cmdList = schedule.command || [];
        const cronExpr = schedule.cron;
        const name = schedule.name || "job";
        const logFile = schedule.log_file;
        const envDict = schedule.env || {};
        const extraArgs = schedule.args || [];

        const fullCmd = [...cmdList, ...extraArgs];

        const lp = logPath(logDir, logFile, name);
        fs.ensureDirSync(path.dirname(lp));

        const envExports = Object.entries(envDict)
            .map(([k, v]) => `${k}='${v}'`)
            .join(' ');

        const cmdStr = fullCmd.join(' ');

        const entry = `${cronExpr} cd ${cwd} && ${envExports ? envExports + ' ' : ''}${cmdStr} >> ${lp} 2>&1`;
        lines.push(entry);
    }
    return lines;
}

function applyCron(lines: string[]) {
    if (lines.length === 0) return;
    const content = lines.join("\n") + "\n";
    console.log("Applying cron:");
    console.log(content);

    spawnSync("crontab", ["-"], {
        input: content,
        encoding: 'utf-8'
    });
}

async function main() {
    const args = process.argv.slice(2);
    const config = loadConfig();
    const automation = config.automation || {};

    if (automation.enabled === false) return;

    if (!args.includes("--skip-services")) {
        await startServices(automation);
    }

    if (args.includes("--skip-cron")) return;

    const lines = buildSchedule(automation);
    if (args.includes("--install-cron")) {
        applyCron(lines);
    } else {
        lines.forEach(l => console.log(l));
    }
}

main().catch(console.error);

File: scripts/automation.py
#!/usr/bin/env python3
# ruff: noqa: I001
from __future__ import annotations

import sys
import argparse
import os
import shlex
import subprocess
from pathlib import Path  # noqa: I001

ROOT = Path(__file__).resolve().parents[1]
sys.path.append(str(ROOT))

from src.config import load_config as load_config_dict  # noqa: E402


def absolute(path: str | None) -> Path:
    if path is None:
        return ROOT
    value = Path(path)
    if value.is_absolute():
        return value
    return ROOT / value


def log_path(base: str, override: str | None, name: str) -> Path:
    if override:
        return absolute(override)
    directory = absolute(base)
    return directory / f"{name}.log"


def merge_env(values: dict[str, str]) -> dict[str, str]:
    data = os.environ.copy()
    for key, value in values.items():
        if value is not None:
            data[key] = value
    return data


def service_command(activate: str | None, cwd: str | None, command: list[str]) -> list[str]:
    parts: list[str] = []
    if activate:
        parts.append(f". {shlex.quote(str(absolute(activate)))}")
    target = shlex.quote(str(absolute(cwd or ".")))
    parts.append(f"cd {target}")
    parts.append(shlex.join(command))
    return ["bash", "-lc", " && ".join(parts)]


def start_services(automation: dict) -> None:
    venv_activate = automation.get("venv_activate")
    log_dir = automation.get("log_dir", "logs/automation")

    for service in automation.get("services", []):
        if not service.get("enabled", True): # Default true if unknown, usually explicit in yaml
            continue

        cwd = service.get("cwd")
        cmd_list = service.get("command", [])
        env_dict = service.get("env", {})
        name = service.get("name", "service")
        log_file = service.get("log_file")
        background = service.get("background", False)

        command = service_command(venv_activate, cwd, cmd_list)
        env = merge_env(env_dict)
        path = log_path(log_dir, log_file, name)
        path.parent.mkdir(parents=True, exist_ok=True)

        with path.open("ab") as stream:
            if background:
                subprocess.Popen(
                    command,
                    stdin=subprocess.DEVNULL,
                    stdout=stream,
                    stderr=stream,
                    env=env,
                    cwd=ROOT,
                    start_new_session=True,
                )
            else:
                subprocess.run(
                    command,
                    stdin=subprocess.DEVNULL,
                    stdout=stream,
                    stderr=stream,
                    env=env,
                    cwd=ROOT,
                    check=True,
                )


def schedule_line(automation: dict, schedule: dict) -> str:
    venv_activate = automation.get("venv_activate")
    log_dir = automation.get("log_dir", "logs/automation")

    cwd = schedule.get("cwd")
    env_dict = schedule.get("env", {})
    cmd_list = schedule.get("command", [])
    cron_expr = schedule.get("cron")
    name = schedule.get("name", "job")
    log_file = schedule.get("log_file")

    parts: list[str] = []
    if venv_activate:
        parts.append(f". {shlex.quote(str(absolute(venv_activate)))}")
    parts.append(f"cd {shlex.quote(str(absolute(cwd or '.')))}")

    env_map = {key: value for key, value in env_dict.items() if value is not None}
    
    # Add support for appending extra arguments defined in yaml
    extra_args = schedule.get("args", [])
    cmd_list.extend(extra_args)

    command = shlex.join(cmd_list)
    if env_map:
        exports = " ".join(f"{key}={shlex.quote(str(value))}" for key, value in env_map.items())
        command = f"{exports} {command}"
    parts.append(command)

    path = log_path(log_dir, log_file, name)
    path.parent.mkdir(parents=True, exist_ok=True)
    return f"{cron_expr} {' && '.join(parts)} >> {shlex.quote(str(path))} 2>&1"


def build_schedule(automation: dict) -> list[str]:
    lines: list[str] = []
    for schedule in automation.get("schedules", []):
        if not schedule.get("enabled", True): # Default to true?
            continue
        lines.append(schedule_line(automation, schedule))
    return lines


def apply_cron(lines: list[str]) -> None:
    if not lines:
        return
    content = "\n".join(lines) + "\n"
    subprocess.run(["crontab", "-"], input=content, text=True, check=True)


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--skip-services", action="store_true")
    parser.add_argument("--skip-cron", action="store_true")
    parser.add_argument("--install-cron", action="store_true")
    return parser.parse_args()


def main() -> None:
    args = parse_args()
    config = load_config_dict()
    automation = config.get("automation", {})

    if not automation.get("enabled", False):
        return

    if not args.skip_services:
        start_services(automation)

    if args.skip_cron:
        return

    lines = build_schedule(automation)
    if args.install_cron:
        apply_cron(lines)
    else:
        for line in lines:
            print(line)


if __name__ == "__main__":
    main()
